src/audio/tdfb/tdfb_ipc4.c

```diff
@@ -0,0 +1,198 @@
+#include <ipc4/header.h>
+#include <ipc4/base-config.h>
+#include <rtos/string.h>
+#include <ipc/stream.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/data_blob.h>
+#include <sof/list.h>
+#include <stdint.h>
+
+#include "tdfb.h"
+#include "tdfb_comp.h"
+
+struct tdfb_notification_payload {
+	struct sof_ipc4_notify_module_data module_data;
+	struct sof_ipc4_control_msg_payload control_msg;
+	struct sof_ipc4_ctrl_value_chan control_value; /* One channel value */
+};
+
+LOG_MODULE_DECLARE(tdfb, CONFIG_SOF_LOG_LEVEL);
+
+static struct ipc_msg *tdfb_notification_init(struct processing_module *mod,
+					      uint32_t control_type_param_id,
+					      uint32_t control_id)
+{
+	struct ipc_msg msg_proto;
+	struct comp_dev *dev = mod->dev;
+	struct comp_ipc_config *ipc_config = &dev->ipc_config;
+	union ipc4_notification_header *primary =
+		(union ipc4_notification_header *)&msg_proto.header;
+	struct ipc_msg *msg;
+	struct tdfb_notification_payload *payload;
+
+	/* Clear header, extension, and other ipc_msg members */
+	memset_s(&msg_proto, sizeof(msg_proto), 0, sizeof(msg_proto));
+	primary->r.notif_type = SOF_IPC4_MODULE_NOTIFICATION;
+	primary->r.type = SOF_IPC4_GLB_NOTIFICATION;
+	primary->r.rsp = SOF_IPC4_MESSAGE_DIR_MSG_REQUEST;
+	primary->r.msg_tgt = SOF_IPC4_MESSAGE_TARGET_FW_GEN_MSG;
+	msg = ipc_msg_w_ext_init(msg_proto.header, msg_proto.extension,
+				 sizeof(struct tdfb_notification_payload));
+	if (!msg)
+		return NULL;
+
+	payload = (struct tdfb_notification_payload *)msg->tx_data;
+	payload->module_data.instance_id = IPC4_INST_ID(ipc_config->id);
+	payload->module_data.module_id = IPC4_MOD_ID(ipc_config->id);
+	payload->module_data.event_id = SOF_IPC4_NOTIFY_MODULE_EVENTID_ALSA_MAGIC_VAL |
+		control_type_param_id;
+	payload->module_data.event_data_size = sizeof(struct sof_ipc4_control_msg_payload) +
+		sizeof(struct sof_ipc4_ctrl_value_chan);
+	payload->control_msg.id = control_id;
+	payload->control_msg.num_elems = 1;
+	payload->control_value.channel = 0;
+
+	comp_dbg(dev, "instance_id = 0x%08x, module_id = 0x%08x",
+		 payload->module_data.instance_id, payload->module_data.module_id);
+	return msg;
+}
+
+static void tdfb_send_notification(struct ipc_msg *msg, uint32_t val)
+{
+	struct tdfb_notification_payload *ipc_payload;
+
+	ipc_payload = (struct tdfb_notification_payload *)msg->tx_data;
+	ipc_payload->control_value.value = val;
+	ipc_msg_send(msg, NULL, false);
+}
+
+int tdfb_common_notification_init(struct processing_module *mod)
+{
+	struct tdfb_comp_data *cd = module_get_private_data(mod);
+
+	cd->msg = tdfb_notification_init(mod, SOF_IPC4_ENUM_CONTROL_PARAM_ID,
+					 CTRL_INDEX_AZIMUTH_ESTIMATE);
+	if (!cd->msg) {
+		comp_err(mod->dev, "Failed to initialize control notification.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+void tdfb_send_notification_common(struct processing_module *mod)
+{
+	struct tdfb_comp_data *cd = module_get_private_data(mod);
+
+	tdfb_send_notification(cd->msg, cd->az_value_estimate);
+}
```

Comment:
two more non-"_common" functions

Reply:
I recall I proposed some functions names change, I wonder what would be better names @lyakh for the layers needed for ipc3 and ipc4 handling?

Reply:
Ah, there was same discussion yesterday. @kv2019i proposed "priv".